Add example for nested PECO (peco-peco)

justapig9020 opened the pull request:
Add an example for nested PECO.
To achieve this, there are some modifications in `task.py` and `main.py`

## task.py
- Fix the bug which causing an exception if there is only one format in the task
- To accept format that matches directories instead of files, add end of string checking into `format_re`
    > Otherwise, the `format_re` may matches one of the file within the directory but the directory itself.
- Add some unit tests for the above new features

## main.py
- Return 1 if any task failed
    > The outer PECO use the returned code to identify whether all of the tasks of the inner PECO are success

justapig9020 committed:
Add example to build up gudge environment

justapig9020 committed:
List not only files but also directories

Since some of the tasks are using directy as its input data, it is
better to list not only files but also directories.

justapig9020 committed:
Matches the file without followed characters

We list the files recursively and in some cases we want to match the
directory. If we allow the matches with followed characters, we may
mathed the file within the directory but not the directory itself.

justapig9020 committed:
Avoid file check if there is only one file type

justapig9020 committed:
Let outter peco clear build env before each task

justapig9020 committed:
Return 1 if some task failed, otherwise 0.

justapig9020 committed:
Use bash in dockerfile to avoid git checkout issue

justapig9020 committed:
Propagate out return code from inner peco

justapig9020 committed:
Remove unnecessary image removement

Since images for each task are the same, we don't have to remove the
image.

justapig9020 committed:
Make the submit file looks faster

justapig9020 committed:
Shhh, let inner peco quiet

To avoid possible injection from submited code, let inner peco run in
quiet mode

justapig9020 committed:
Add make clean for example/nested

justapig9020 committed:
Update directory structure in yaml's comment

1am9trash commented on example/nested/outter.yaml:
The `Makefile` cannot run well in my environment (wsl2 with ubuntu 22.04). I try to figure out the issue, and it seems like wsl mounted filesystem problem. The `docker-compose up` may somehow change the working context, and then fail the following commands. You may find that `getcwd()` error in picture.

![螢幕擷取畫面 2023-09-16 014712](https://github.com/ntu-eclab-ca-ta/PECO/assets/69944614/08d0a968-a248-4454-834b-ba88fcd046d4)

1am9trash replied:
It can run on my ubuntu environment in vm successfully. I also try to use a docker inside docker to handle environment problem, but it turns out a mess, so I just stop the attemp.

justapig9020 replied:
Ok, I will reproduce the issue on WSL env and try to fix it.

1am9trash replied:
Just let you know, I think it's ok to ignore this issue if you want. After testing it, it may be a wsl own problem with uncertainly support of docker, not really PECO repo problem.

justapig9020 replied:
We should make sure that `example/local`, which is tend to run on student side, works will for all environments, including `linux`, `windows`, and `mac`.
For `peco-peco`, since it will only be run on our server, I agree that it's OK to ignore the bugs on windows for now.

justapig9020 replied:
I think we can create on some github actions for these testings in the near future

1am9trash commented on example/nested/private/dockerfile:
> `apt-get` may be considered as lower-level and "back-end", and support other APT-based tools. `apt` is designed for end-users (human) and its output may be changed between versions.

When using dockerfile, find that `apt install` may cause error and `apt does not have a stable cli interface.` warning. Different with `apt-get`, `apt` is a human-friendly tool having more interaction with user. It might fail the script sometimes.

So maybe we should use `apt-get` in dockerfile instead.

justapig9020 replied:
Wow, I didn't know this. I think you can create a new PR for this.

1am9trash replied:
Ok, I may deal with this later.

1am9trash commented on src/task.py:
```diff
@@ -68,4 +76,5 @@ def list_matched_files(files, format):
     format = re.compile(format)
     files = [file for file in files]
```
I think `line 77` is redundant.

justapig9020 replied:
I have removed the redundant code. Thx.

1am9trash commented on src/task.py:
```diff
@@ -68,4 +76,5 @@ def list_matched_files(files, format):
     files = [file for file in files]
     indexed_files = [(file, format.findall(file)) for file in files]
+    # TODO: Maybe we should raise an error if there are multiple matches
     return [IndexedFile(index[0], file) for file, index in indexed_files if index != []]
```
The comment sounds reasonable. I think it may like below code.

```python
class AmbigousIndex(Exception):
    def __init__(self, filename, match_cnt, format_re):
        super().__init__(f'File "{filename}" has "{match_cnt}" matches with format regex "{format_re}"')

def list_matched_files(files, format):
    # Filter the files
    pattern = re.compile(format)
    indexed_files = []
    for file in files:
        match = re.findall(pattern, file)
        if len(match) > 1:
            raise AmbigousIndex(file, len(match), pattern)
        elif len(match) == 1:
            indexed_files.append(IndexedFile(match[0], file))
    return indexed_files
```

justapig9020 replied:
I have add some comments for this issue. I think under the current implementation, the ambiguous index will never happen. So we don't need to deal with it.

1am9trash replied:
I think this is decided by `index regex` design. Now we have index like `[0-9]+`, so there won't be mistakes. If we are sure that `index regex` formats are clean, I agree we don't need to handle it.

justapig9020 replied:
I think we don't have the responsibility and ability to distinguish the wrong `index regex` from user. Therefore, we don't need to handle it.

1am9trash commented on example/nested/outter.yaml:
```diff
@@ -0,0 +1,37 @@
+# └── submit
+#     └── <id>
+#         └── src
+#             └── <submited code>
```
Is this draw with `tree` command?

justapig9020 replied:
This was draw with `tree` command and been manually modified

1am9trash commented:
> Add some unit tests for the above new features

The comment said new unit tests is added, but I didn't see any change in test directory. Am I misunderstanding?

justapig9020 commented:
I have wrote some tests but I have no idea why they disappeared. I will rewrite them later. I am sorry for that.

justapig9020 committed:
Remove redundant code

justapig9020 committed:
Add some comments to explain why ambigous index never happens

justapig9020 committed:
Add unit test for "input_only"

The PECO should be able to match any number of formats

justapig9020 committed:
Add unit test for "match directories"

Since we allow task format to match directories, here is the unit test
for the feature

justapig9020 commented:
Sorry for the late reply. I have updated the unit tests for the new features.
Including
- Variable format count (one in the unit test)
- Matches directories
